File: src/cli/src/distill.rs
//! D-Distill:按主题聚合日志条目 → materials/<topic>.md(纯拼接)。

use std::fs;
use std::path::{Path, PathBuf};

use crate::frontmatter::FrontMatter;
use crate::journal::{self, Entry};

/// 执行 distill:聚合归属 `topic` 的条目,生成 materials/<topic>.md。
/// 文件已存在时重新生成(幂等)。返回输出路径。
pub fn distill(workdir: &Path, topic: &str) -> Result<PathBuf, String> {
    let journals = journal::read_all(workdir)?;

    // 按 YAML 归属收集条目
    let mut collected: Vec<Entry> = vec![];
    for jf in &journals {
        for (id, t) in &jf.fm.topics {
            if t == topic {
                if let Some(e) = jf.entries.iter().find(|e| &e.id == id) {
                    collected.push(e.clone());
                }
            }
        }
    }
    if collected.is_empty() {
        return Err(format!("主题「{}」暂无归属条目,先用 material organize 提取主题", topic));
    }

    // 按文件 + 时间排序
    collected.sort_by(|a, b| (a.file.clone(), a.id.clone()).cmp(&(b.file.clone(), b.id.clone())));

    // 素材 front matter
    let mut fm = FrontMatter::default();
    fm.topic = Some(topic.to_string());
    for e in &collected {
        fm.sources.push(format!("journal/{}", e.reference()));
    }

    // 正文:纯聚合拼接,每条带来源引用
    let mut body = String::new();
    for e in &collected {
        if !body.is_empty() {
            body.push('\n');
        }
        body.push_str(&e.text);
        body.push_str(&format!("\n\n> 来源:journal/{}\n", e.reference()));
    }

    let dir = workdir.join("materials");
    fs::create_dir_all(&dir).map_err(|e| format!("mkdir materials: {e}"))?;
    let path = dir.join(format!("{}.md", topic));
    fs::write(&path, format!("{}{}", crate::frontmatter::render(&fm), body))
        .map_err(|e| format!("write {}: {e}", path.display()))?;
    Ok(path)
}

/// 执行 distill 提炼:先聚合,再调 LLM 删除次要信息,输出提炼稿
/// `materials/<topic>-refined.md`(聚合文件保留,人工可对比)。
pub fn distill_refine(workdir: &Path, topic: &str) -> Result<PathBuf, String> {
    let base = distill(workdir, topic)?;
    let material = fs::read_to_string(&base).map_err(|e| format!("读聚合稿: {e}"))?;
    let prompt = format!(
        "以下是从日志聚合的写作素材(主题:{})。\n\
         请删除次要信息、保留要点,输出提炼后的素材(markdown,与素材同语言)。\n\
         只输出提炼结果,不要解释。\n\n素材:\n{}",
        topic, material
    );
    let refined = crate::organize::run_llm(&prompt)?;
    let path = workdir.join("materials").join(format!("{}-refined.md", topic));
    fs::write(&path, refined).map_err(|e| format!("write {}: {e}", path.display()))?;
    Ok(path)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn tmpdir(name: &str) -> PathBuf {
        let dir = std::env::temp_dir().join(format!("material-distill-{}-{}", name, std::process::id()));
        let _ = fs::remove_dir_all(&dir);
        fs::create_dir_all(&dir).unwrap();
        dir
    }

    #[test]
    fn distill_aggregates_by_topic() {
        let dir = tmpdir("agg");
        let jdir = dir.join("journal");
        fs::create_dir_all(&jdir).unwrap();
        // 2026-08-15.md:两条条目,一条归宣传册
        fs::write(
            jdir.join("2026-08-15.md"),
            "---\ntopics:\n  \"09-30\": 宣传册\n  \"12-15\": 内容策略\n---\n\n## 09:30\n宣传册想法一\n\n## 12:15\n内容策略想法\n",
        )
        .unwrap();
        // 2026-08-16.md:一条归宣传册
        fs::write(
            jdir.join("2026-08-16.md"),
            "---\ntopics:\n  \"08-00\": 宣传册\n---\n\n## 08:00\n宣传册想法二\n",
        )
        .unwrap();

        let out = distill(&dir, "宣传册").unwrap();
        let doc = fs::read_to_string(&out).unwrap();
        assert!(doc.contains("topic: 宣传册"));
        assert!(doc.contains("journal/2026-08-15.md#09-30"));
        assert!(doc.contains("journal/2026-08-16.md#08-00"));
        assert!(doc.contains("宣传册想法一"));
        assert!(doc.contains("宣传册想法二"));
        assert!(!doc.contains("内容策略想法"));
        let _ = fs::remove_dir_all(&dir);
    }

    #[test]
    fn distill_empty_topic_errors() {
        let dir = tmpdir("empty");
        let out = distill(&dir, "不存在");
        assert!(out.is_err());
        let _ = fs::remove_dir_all(&dir);
    }
}

File: src/studio/lib/workflow/distill.dart
/// D-Distill（提取）：读分组产物，过滤次要信息并统一表达，
/// 形成初稿 materials/<主题>.md（产物 03，与 CLI distill.rs 对齐）。
///
/// 产物链：01 收集（collect）→ 02 分组（organize）→ 03 初稿（distill）→ 04 定稿（express）。
library;

import 'dart:io';

import 'package:path/path.dart' as p;

import 'frontmatter.dart';

/// 初稿提示词：过滤次要信息 + 成文（通用判据，不针对特定文本）。
String buildDraftPrompt(String topic, String aggregate) {
  return '以下是从日志中按主题聚合的写作素材(主题:$topic)。请基于素材形成一篇初稿:\n'
      '1. 过滤次要信息——从严删除,仅限以下四类杂质:\n'
      '- 过程性叙述:记录"正在做/打算做"的过程性语句(如"我打算""我在尝试""最近"),只保留其中的结论或洞察\n'
      '- 用途信息:关于素材将被用于什么、发布到哪里的说明(如发布渠道、宣传用途、收益安排、目标账号),除非该用途本身是主题的核心问题\n'
      '- 重复表述:同一观点多次出现时只保留最完整的一次\n'
      '- 即时情绪宣泄:直接的情绪感叹与自我对话;若其中包含可迁移的洞察则保留洞察\n'
      '2. 保留从宽:与主题相关的信息细节(人物、事件、设定、场景、动机、洞察、金句)全部保留,信息宁可完整不可削减\n'
      '3. 成文——不是素材的顺述或改写,而是基于素材的完整作品:\n'
      '- 提炼立意:这篇文章想表达什么(素材里反复出现、最能打动人心的内核)\n'
      '- 组织结构:按立意层层推进,有开头、发展、收束,主题分层展开\n'
      '- 统一视角与语气,语言有节奏,金句自然浮现\n'
      '输出一篇结构完整、可独立阅读的初稿(markdown,与素材同语言)。只输出初稿,不要解释。\n\n素材:\n$aggregate';
}

/// 执行 distill：读分组产物 groups/<主题>.md → LLM 过滤次要信息并统一表达
/// → 初稿 materials/<主题>.md（产物 03）。返回产物路径。
/// [llm] 为 LLM 回调（接收 prompt 返回文本），便于测试注入。
Future<String> distill(
  String workdir,
  String topic,
  Future<String> Function(String prompt) llm,
) async {
  final src = p.join(workdir, 'groups', '$topic.md');
  final groupFile = File(src);
  if (!groupFile.existsSync()) {
    throw Exception('读分组 $src: 不存在(先运行 organize 生成分组)');
  }
  final group = groupFile.readAsStringSync();
  final (fm, body) = FrontMatter.parse(group);

  final prompt = buildDraftPrompt(topic, body);
  final draft = await llm(prompt);

  // front matter：沿用分组的 sources
  final f = fm ?? FrontMatter();
  f.topic = topic;

  final dir = p.join(workdir, 'materials');
  Directory(dir).createSync(recursive: true);
  final path = p.join(dir, '$topic.md');
  File(path).writeAsStringSync('${f.render()}$draft');
  return path;
}
